toonz/sources/toonz/filmstripcommand.cpp

```diff
@@ -53,9 +53,58 @@ TFrameId operator+(const TFrameId &fid, int d) {
 
 //-----------------------------------------------------------------------------
 
+void updateSubXSheet(TXshSimpleLevel *sl, std::vector<TFrameId> oldFids,
+                     std::vector<TFrameId> newFids, TXsheet *xsh,
+                     std::vector<TXshChildLevel *> *childLevels) {
+  for (int c = 0; c < xsh->getColumnCount(); ++c) {
+    int r0, r1;
+    int n = xsh->getCellRange(c, r0, r1);
+    if (n > 0) {
+      bool changed = false;
+      std::vector<TXshCell> cells(n);
+      xsh->getCells(r0, c, n, &cells[0]);
+      for (int i = 0; i < n; i++) {
+        TXshCell currCell = cells[i];
+        // check the sub xsheets too
+        if (!cells[i].isEmpty() &&
+            cells[i].m_level->getType() == CHILD_XSHLEVEL) {
+          TXshChildLevel *level = cells[i].m_level->getChildLevel();
+          // make sure we haven't already checked the level
+          if (level &&
+              std::find(childLevels->begin(), childLevels->end(), level) ==
+                  childLevels->end()) {
+            childLevels->push_back(level);
+            TXsheet *subXsh = level->getXsheet();
+            updateSubXSheet(sl, oldFids, newFids, subXsh, childLevels);
+          }
+        }
+        for (int j = 0; j < oldFids.size(); j++) {
+          if (oldFids.at(j) == newFids.at(j)) continue;
+          TXshCell tempCell(sl, oldFids.at(j));
+          bool sameSl  = tempCell.getSimpleLevel() == currCell.getSimpleLevel();
+          bool sameFid = tempCell.getFrameId() == currCell.getFrameId();
+          if (sameSl && sameFid) {
+            TXshCell newCell(sl, newFids.at(j));
+            cells[i] = newCell;
+            changed  = true;
+            break;
+          }
+        }
+      }
+      if (changed) {
+        xsh->setCells(r0, c, n, &cells[0]);
+        TApp::instance()->getCurrentXsheet()->notifyXsheetChanged();
+      }
+    }
+  }
+}
+
+//-----------------------------------------------------------------------------
+
 void updateXSheet(TXshSimpleLevel *sl, std::vector<TFrameId> oldFids,
                   std::vector<TFrameId> newFids, TXsheet *xsh) {
```

Comment:
In your latest modification, the function `updateXsheet()` will not be called recursively anymore. 
Now it is always called after `TXsheet *xsh = TApp::instance()->getCurrentScene()->getScene()->getTopXsheet();` , right? 
If so, IMO the argument `TXsheet *xsh` can be removed and obtaining it inside the function would be clear.

Reply:
Good catch.  I updated this.